pypeit/core/coadd2d.py

```diff
@@ -0,0 +1,596 @@
+    snr_bar = np.zeros(nexp)
+    if echelle:
+        norders = specobjs_list[0].ech_orderindx.max() + 1
+        for iexp, sobjs in enumerate(specobjs_list):
+            uni_objid = np.unique(sobjs.ech_objid)
+            nobjs = len(uni_objid)
+            order_snr = np.zeros((norders, nobjs))
+            for iord in range(norders):
+                for iobj in range(nobjs):
+                    ind = (sobjs.ech_orderindx == iord) & (sobjs.ech_objid == uni_objid[iobj])
+                    flux = sobjs[ind][0].optimal['COUNTS']
+                    sig = sobjs[ind][0].optimal['COUNTS_SIG']
+                    wave = sobjs[ind][0].optimal['WAVE']
+                    mask = sobjs[ind][0].optimal['MASK']
+                    rms_sn, weights = coadd.sn_weights(flux, sig, mask, wave, const_weights=True)
+                    order_snr[iord, iobj] = rms_sn
+
+            # Compute the average SNR and find the brightest object
+            snr_bar_vec = np.mean(order_snr, axis=0)
+            objid[iexp] = uni_objid[snr_bar_vec.argmax()]
+            snr_bar[iexp] = snr_bar_vec[snr_bar_vec.argmax()]
+    else:
+        msgs.error('Automated objid selection is not yet implemented for multislit')
+        # -- for multislit this routine will:
+        #  Determine which object on the slitmask is brightest and attempt to match them up somehow.  The problem is
+        #  how to associate the objects together on the slits if they have moved.
+
+    return objid, snr_bar
+
+def optimal_weights(specobjs_list, slitid, objid, echelle=True):
+
+    nexp = len(specobjs_list)
+    nspec = specobjs_list[0][0].trace_spat.shape[0]
+    # Grab the traces, flux, wavelength and noise for this slit and objid.
+    trace_stack = np.zeros((nexp, nspec), dtype=float)
+    flux_stack = np.zeros((nexp, nspec), dtype=float)
+    sig_stack = np.zeros((nexp, nspec), dtype=float)
+    wave_stack = np.zeros((nexp, nspec), dtype=float)
+    mask_stack = np.zeros((nexp, nspec), dtype=bool)
+
+    for iexp, sobjs in enumerate(specobjs_list):
+        ithis = (sobjs.slitid == slitid) & (sobjs.objid == objid[iexp])
+        trace_stack[iexp, :] = sobjs[ithis].trace_spat
+        flux_stack[iexp,:] = sobjs[ithis][0].optimal['COUNTS']
+        sig_stack[iexp,:] = sobjs[ithis][0].optimal['COUNTS_SIG']
+        wave_stack[iexp,:] = sobjs[ithis][0].optimal['WAVE']
+        mask_stack[iexp,:] = sobjs[ithis][0].optimal['MASK']
+
+    # TODO For now just use the zero as the reference for the wavelengths? Perhaps we should be rebinning the data though?
+    rms_sn, weights = coadd.sn_weights(flux_stack, sig_stack, mask_stack, wave_stack)
+
+    return rms_sn, weights, trace_stack, wave_stack
+
+def load_coadd2d_stacks(spec2d_files):
+
+    # Grab the files
+    spec1d_files = [files.replace('spec2d', 'spec1d') for files in spec2d_files]
+    # Get the master dir
+    head0 = fits.getheader(spec2d_files[0])
+    mdir = os.path.basename(head0['PYPMFDIR'])+'/'
```

Comment:
I wonder if this should be a method somewhere so that 
we don't have to hard-code the PYPMFDIR card in more
than one place.